### PR TITLE
Add @propagate_inbounds to getindex for SMatrix

### PR DESCRIPTION
**Pull request by rafaqz**
Needed to work with GPUArrays. See: https://github.com/JuliaGPU/GPUArrays.jl/issues/157

### CODE REVIEW

File: src/SMatrix.jl [hunk 1]
```diff
@@ -71,7 +71,7 @@ end
 ## SMatrix methods ##
 #####################
 
-function getindex(v::SMatrix, i::Int)
+Base.@propagate_inbounds function getindex(v::SMatrix, i::Int)
     Base.@_inline_meta
```
[the remark this answers is not in the document]
Comment (SimonDanisch) [reply src/SMatrix.jl]: ah is that true? so you dont need the inline meta?

[the remark this answers is not in the document]
Comment (mschauer) [reply src/SMatrix.jl]:
```
  @propagate_inbounds

  Tells the compiler to inline a function while retaining the caller's
  inbounds context.
```

[the remark this answers is not in the document]
Comment (SimonDanisch) [reply src/SMatrix.jl]: haha, guilty! never read the docs!

[the remark this answers is not in the document]
Comment (rafaqz) [reply src/SMatrix.jl]: Base is full of @inline @propagate_inbounds so maybe you're not the only one

[the remark this answers is not in the document]
Comment (SimonDanisch) [reply src/SMatrix.jl]: yeah, thats were i got it from ;)

**Comment by andyferris** [comment]
Thanks!

Can we do this properly for all the array types?

**Comment by rafaqz** [comment]
I'm not sure the inlining comments were meant as no-go for `@propagate_inbounds` or not?

SMatrix is especially useful on the GPU, it's the only type of matrix you can use as a struct field. But I can update this for everything else too if that's a good idea.

**Comment by tkoolen** [comment]
> I'm not sure the inlining comments were meant as no-go for @propagate_inbounds or not?

Nope, just wanted to point out redundant code.

**Comment by andyferris** [comment]
As in: keep the propagate inbounds, ditch the inline.

**Comment by andyferris** [comment]
My point was partly that SMatrix is only one type of SArray (does the same problem occur for SVector? Or 3D arrays?)

**Comment by rafaqz** [comment]
Ok duh I missed the `Base.@_inline_meta`

I'll test all types of SArray and FieldVector and update.

**Comment by rafaqz** [comment]
SArrays needed `@propagate_inbounds` to work, SVector already has it and also works. Scalars work without it because of constant propagation.

FieldVectors also have `@propagate_inbounds` already it but don't work on GPUs for some other reason I don't understand, even when the underlying type is immutable.

Is there anything else relevant to the (immutable) GPU context?

**Comment by rafaqz** [comment]
Also I haven't added any tests for this because of the need for a GPU. 

Is it possible to test GPUArrays integration with JLArray? @SimonDanisch ?

**Comment by SimonDanisch** [comment]
I mean, all of this isn't GPU specific! You will want to have inbounds propagating in all performance sensitive code. Not sure how to test this. A hacky way would be to search for the throw in the IR :P

**Comment by rafaqz** [comment]
Not sure I can manage that kind of hack! And sure its useful for other things as well, but it doesn't break them entirely without it! I just meant having some test to ensure
these types continued to actually work on the GPU.

**Pull request merged**